### PR TITLE
Calendars plugin: Bahai and Chinese (plus variants)

### PR DESCRIPTION
**Pull request by gzotti**
### Description

This brings 6 new calendars into the plugin. 
The Chinese calendar had been started with #1339, but we had decided this was not fit for use. 

Fixes #1339 (issue)

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] This change requires a documentation update

### How Has This Been Tested?

I added unit tests where applicable. The Chinese calendar can only be tested via scripting. Tests run OK. 

**Test Configuration**:
* Operating system: Windows 10
* Graphics Card: (irrelevant)

## Checklist:

- [x] My code follows the [code style](http://stellarium.org/doc/head/codingStyle.html) of this project.
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (header file)
- [x] I have updated the respective chapter in the Stellarium User Guide
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by alex-w** [comment]
Please see screenshot:
stellarium-013

Why Chinese tab is empty?

The info line for new calendars is too long - could it be more compact?

**Comment by 10110111** [comment]
@alex-w your screenshot has pixelized bottom panel icons — is it the current version + this PR? If so, looks like HighDPI support is broken on your system (which should be treated in a separate issue).

**Comment by alex-w** [comment]
> @alex-w your screenshot has pixelized bottom panel icons — is it the current version + this PR?

Yes

> If so, looks like HighDPI support is broken on your system (which should be treated in a separate issue).

PNG:
stellarium-002

**Comment by gzotti** [comment]
grafik
Oops. Why do I never think of horizontal screenspace issues... 0:-)

The Chinese tab is visible only in non-Release builds (has been like that for a year now...). I hide it in the Release builds, but it will show elements in future extensions.

**Comment by alex-w** [comment]
> I didn't intend to hijack this thread, could you create a new issue?

Yes, but I see no pixelation effect for bottom panel icons in real hardware - only on the screenshots

**Comment by gzotti** [comment]
Note that the screenshots are set up separately from regular display updates. Maybe make a OS screen capture to see the difference. But again, this belongs to a different issue.

**Review by alex-w — APPROVED** [review]

File: plugins/Calendars/src/JapaneseCalendar.cpp [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Suite 500, Boston, MA  02110-1335, USA.
+ */
+
+#include "StelTranslator.hpp"
+#include "JapaneseCalendar.hpp"
+//#include "GregorianCalendar.hpp"
```
Comment (10110111) [inline plugins/Calendars/src/JapaneseCalendar.cpp:21]: Why is this commented out? I get compilation failures unless I uncomment this line and similar ones in two other files (and I wonder why the CI builds don't fail).

Comment (gzotti) [reply plugins/Calendars/src/JapaneseCalendar.cpp:21]: interesting. The Clangd syntax checker in QtCreator recommended to remove the include as it is not used.
What is the actual error?

Comment (10110111) [reply plugins/Calendars/src/JapaneseCalendar.cpp:21]:
```
/home/ruslan/src/stellarium/plugins/Calendars/src/JapaneseCalendar.cpp:201:17: error: ‘GregorianCalendar’ has not been declared
  201 |  const int year=GregorianCalendar::gregorianYearFromFixed(qRound(floor(rd_t)));
      |                 ^~~~~~~~~~~~~~~~~
make[2]: *** [plugins/Calendars/src/CMakeFiles/Calendars-static.dir/build.make:583: plugins/Calendars/src/CMakeFiles/Calendars-static.dir/JapaneseCalendar.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
/home/ruslan/src/stellarium/plugins/Calendars/src/VietnameseCalendar.cpp: In static member function ‘static StelLocation VietnameseCalendar::vietnameseLocation(double)’:
/home/ruslan/src/stellarium/plugins/Calendars/src/VietnameseCalendar.cpp:243:17: error: ‘GregorianCalendar’ has not been declared
  243 |  const int year=GregorianCalendar::gregorianYearFromFixed(qRound(floor(rd_t)));
      |                 ^~~~~~~~~~~~~~~~~
make[2]: *** [plugins/Calendars/src/CMakeFiles/Calendars-static.dir/build.make:609: plugins/Calendars/src/CMakeFiles/Calendars-static.dir/VietnameseCalendar.cpp.o] Error 1
/home/ruslan/src/stellarium/plugins/Calendars/src/KoreanCalendar.cpp: In static member function ‘static StelLocation KoreanCalendar::koreanLocation(double)’:
/home/ruslan/src/stellarium/plugins/Calendars/src/KoreanCalendar.cpp:259:11: error: ‘GregorianCalendar’ has not been declared
  259 |  if (rd_t<GregorianCalendar::fixedFromGregorian({1908, GregorianCalendar::april, 1}))
      |           ^~~~~~~~~~~~~~~~~
/home/ruslan/src/stellarium/plugins/Calendars/src/KoreanCalendar.cpp:259:56: error: ‘GregorianCalendar’ has not been declared
  259 |  if (rd_t<GregorianCalendar::fixedFromGregorian({1908, GregorianCalendar::april, 1}))
      |                                                        ^~~~~~~~~~~~~~~~~
/home/ruslan/src/stellarium/plugins/Calendars/src/KoreanCalendar.cpp:261:17: error: ‘GregorianCalendar’ has not been declared
  261 |  else if ((rd_t<GregorianCalendar::fixedFromGregorian({1912, GregorianCalendar::january, 1})) ||
      |                 ^~~~~~~~~~~~~~~~~
/home/ruslan/src/stellarium/plugins/Calendars/src/KoreanCalendar.cpp:261:62: error: ‘GregorianCalendar’ has not been declared
  261 |  else if ((rd_t<GregorianCalendar::fixedFromGregorian({1912, GregorianCalendar::january, 1})) ||
      |                                                              ^~~~~~~~~~~~~~~~~
/home/ruslan/src/stellarium/plugins/Calendars/src/KoreanCalendar.cpp:262:12: error: ‘GregorianCalendar’ has not been declared
  262 |    ((rd_t>=GregorianCalendar::fixedFromGregorian({1954, GregorianCalendar::march, 21})) &&
      |            ^~~~~~~~~~~~~~~~~
/home/ruslan/src/stellarium/plugins/Calendars/src/KoreanCalendar.cpp:262:57: error: ‘GregorianCalendar’ has not been declared
  262 |    ((rd_t>=GregorianCalendar::fixedFromGregorian({1954, GregorianCalendar::march, 21})) &&
      |                                                         ^~~~~~~~~~~~~~~~~
/home/ruslan/src/stellarium/plugins/Calendars/src/KoreanCalendar.cpp:263:11: error: ‘GregorianCalendar’ has not been declared
  263 |     (rd_t<GregorianCalendar::fixedFromGregorian({1961, GregorianCalendar::august, 10}))))
      |           ^~~~~~~~~~~~~~~~~
/home/ruslan/src/stellarium/plugins/Calendars/src/KoreanCalendar.cpp:263:56: error: ‘GregorianCalendar’ has not been declared
  263 |     (rd_t<GregorianCalendar::fixedFromGregorian({1961, GregorianCalendar::august, 10}))))
      |                                                        ^~~~~~~~~~~~~~~~~
/home/ruslan/src/stellarium/plugins/Calendars/src/KoreanCalendar.cpp: In static member function ‘static int KoreanCalendar::koreanNewYearInGregorianYear(int)’:
/home/ruslan/src/stellarium/plugins/Calendars/src/KoreanCalendar.cpp:446:17: error: ‘GregorianCalendar’ has not been declared
  446 |  return gYear+1-GregorianCalendar::gregorianYearFromFixed(ChineseCalendar::chineseEpoch)-304;
      |                 ^~~~~~~~~~~~~~~~~
/home/ruslan/src/stellarium/plugins/Calendars/src/KoreanCalendar.cpp: In static member function ‘static int KoreanCalendar::koreanAge(QVector<int>, int)’:
/home/ruslan/src/stellarium/plugins/Calendars/src/KoreanCalendar.cpp:470:20: error: ‘GregorianCalendar’ has not been declared
  470 |  const int gYearRD=GregorianCalendar::gregorianFromFixed(rd).at(0);
      |                    ^~~~~~~~~~~~~~~~~
/home/ruslan/src/stellarium/plugins/Calendars/src/KoreanCalendar.cpp:472:23: error: ‘GregorianCalendar’ has not been declared
  472 |  const int gYearBirth=GregorianCalendar::gregorianFromFixed(rdBirth).at(0);
      |                       ^~~~~~~~~~~~~~~~~
```

Comment (gzotti) [reply plugins/Calendars/src/JapaneseCalendar.cpp:21]: Ah, I use the Precompiled Headers feature and have Gregorian included in the PCH list. 
You are right, these headers are required in this case. Fixed in master.

**Pull request merged**